### PR TITLE
Ignore case on column alias names from subquery

### LINKED ISSUE #202: Alias and column names from subquery are case sensitive when used as conditions in outer query
It seem like case must match on alias/column names from subquery when they are used as a condition in the outer query, *only* if the subquery alias is used in the condition.

for example, this works:
```sql
select tEsT
from (select a as TeSt from nowhere) as t
where tEsT = 'asdf'
```

but this fails:
```sql
select tEsT
from (select a as TeSt from nowhere) as t
where t.tEsT = 'asdf'
```
with an error
```error
ValueError: 'tEsT' is not in list
```

I think it can be fixed by ignoring case when checking column alias.

In Parser.py, `_resolve_nested_query, we can change
```py
if column_name in subparser.columns_aliases_names:
```
to
```py
if column_name.lower() in [x.lower() for x in subparser.columns_aliases_names]:
```
or 
```py
if any(filter(lambda x: x.lower() == column_name.lower(),subparser.columns_aliases_names)):
```
... not sure of a nicer way to say it.

 I'll put a pull request w/  a test.

### PR DESCRIPTION
**Pull request by christopherpickering**
Added code change and test.

Resolves #202

### CODE REVIEW

**Comment by macbre** [comment]
@christopherpickering , thanks for your pull request. Please format the code using `black`.

I'll review your change tomorrow.

**Comment by christopherpickering** [comment]
I already found a problem with this... maybe you have an idea!
When checking if a subquery is a rank I used the "get_nth_previous" to count backwards to rank.

It seems like you can have multiple order by columns! Which makes the length from the alias back to RANK unknown!

Is there another command that can get us back from the alias? 

maybe using a combination of `last_keyword_normalized` and if we can get back to ORDER BY, we should be able to have a fixed count back to rank?

There are multiple ways to rank:

```sql
RANK() OVER (ORDER BY Rate DESC) AS RankBySalary  
RANK() OVER (PARTITION BY i.LocationID ORDER BY i.Quantity DESC) AS Rank
RANK() OVER (PARTITION BY SalesTerritoryRegion ORDER BY SUM(SalesAmountQuota) DESC ) AS RankResult  
```
Again, the problem is with "AS" not being required. W/ the AS there, it works. But without the "AS" we need to be able to determine if we are on a rank.

**Comment by collerek** [comment]
@christopherpickering I changed the implementation to treat `over` clause as a new type of "subquery" that simplifies the implementation and makes it independent from a number of order_by columns.

I include the columns used in order_by in over clause in alias definition, but they are treated as select parts of the expression as they do not impose the order on the result rows. (Meaning that in tests the order_by part of columns_dict is empty).

@macbre - now I think it's ready for a review

**Comment by christopherpickering** [comment]
Nice! Thanks!

File: test/test_mssql_server.py [hunk 1]
```diff
@@ -27,38 +27,38 @@ def test_sql_server_cte():
...
-                    WITH foo AS (
-                         SELECT * FROM tab
-                    )
-                    DELETE FROM z JOIN foo ON z.a = foo.a
-                        """.strip()
+                        WITH foo AS (
+                             SELECT * FROM tab
+                        )
+                        DELETE FROM z JOIN foo ON z.a = foo.a
+                            """.strip()
```
Comment (macbre) [inline test/test_mssql_server.py:61]: @collerek - I believe your IDE keeps adding these tabs here on each commit 😉

Comment (collerek) [reply test/test_mssql_server.py:61]: Sneaky PyCharm! 🤣

Comment (collerek) [reply test/test_mssql_server.py]: ok, I think black didn't like the `strip()` which was unnecessary anyway

File: sql_metadata/parser.py [hunk 2]
```diff
@@ -66,7 +66,7 @@ def __init__(self, sql: str = "") -> None:
         self._subquery_level = 0
         self._nested_level = 0
         self._parenthesis_level = 0
-        self._open_parentheses = []
+        self._open_parentheses: List[SQLToken] = []
```
Comment (macbre) [inline sql_metadata/parser.py:69]: Type hinting 👍🏻

**Review by macbre — APPROVED** [review]

**Comment by macbre** [comment]
@collerek - feel free to merge it.

**Pull request merged**